server/api/db/sqldb/db.py

```diff
@@ -1321,8 +1316,31 @@ def _find_artifacts(
             query = self._add_artifact_category_query(category, query)
         if most_recent:
             query = self._attach_most_recent_artifact_query(session, query)
+        if limit:
+            query = query.limit(limit)
+
+        # compile the outer query which returns the tuple of artifact and its tag name
+        # it will join on the subquery matching the artifact IDs
+        # this way, the limit will take effect and optimized
```

Comment:
I feel this comment doesn't really explain what exactly we are doing and why we're doing it
```suggestion
        # compile the above query as a sub query only for filtering out the relevant ids so the 
        # limit operation which sadly needs to load all the results before limiting will work on a slimmer data set
        # and will be much more optimized. Then join the outer query on the subquery to select the correct 
        # columns of the table
```